core/src/main/java/org/apache/brooklyn/core/config/internal/AbstractConfigMapImpl.java

```diff
@@ -107,4 +252,227 @@ public int size() {
+            }
+        };
+        ConfigKey<T> ownKey = keyFn.apply(getContainer());
+        if (ownKey==null) ownKey = queryKey;
+        @SuppressWarnings("unchecked")
+        final Class<T> type = (Class<T>) ownKey.getType();
+        
+        // takes type of own key (or query key if own key not available)
+        // takes default of own key if available and has default, else of query key
+        
+        Function<Maybe<Object>, Maybe<T>> coerceFn = new Function<Maybe<Object>, Maybe<T>>() {
+            @SuppressWarnings("unchecked") @Override public Maybe<T> apply(Maybe<Object> input) {
+                if (raw || input==null || input.isAbsent()) return (Maybe<T>)input;
+                return Maybe.ofAllowingNull(coerce(input.get(), type));
+            }
+        };
+        // prefer default and type of ownKey
+        Maybe<T> defaultValue = raw ? Maybe.<T>absent() :
+            ownKey.hasDefaultValue() ? coerceFn.apply(Maybe.of((Object)ownKey.getDefaultValue())) : 
+            queryKey.hasDefaultValue() ? coerceFn.apply(Maybe.of((Object)queryKey.getDefaultValue())) :
+                Maybe.<T>absent();
+            
+        if (ownKey instanceof ConfigKeySelfExtracting) {
+            
+            Function<TContainer, Maybe<Object>> lookupFn = new Function<TContainer, Maybe<Object>>() {
+                @Override public Maybe<Object> apply(TContainer input) {
+                    Maybe<Object> result = getRawValueAtContainer(input, queryKey);
+                    if (!raw) result = resolveRawValueFromContainer(input, queryKey, result);
+                    return result;
+                }
+            };
+            Function<TContainer, TContainer> parentFn = new Function<TContainer, TContainer>() {
+                @Override public TContainer apply(TContainer input) {
+                    return getParentOfContainer(input);
+                }
+            };
+            AncestorContainerAndKeyValueIterator<TContainer, T> ckvi = new AncestorContainerAndKeyValueIterator<TContainer,T>(
+                getContainer(), keyFn, lookupFn, coerceFn, parentFn);
+            
+            return ConfigInheritances.resolveInheriting(
+                getContainer(), ownKey, coerceFn.apply(lookupFn.apply(getContainer())), defaultValue, 
+                ckvi, InheritanceContext.RUNTIME_MANAGEMENT, getDefaultRuntimeInheritance());
+            
+        } else {
+            String message = "Config key "+ownKey+" of "+getBrooklynObject()+" is not a ConfigKeySelfExtracting; cannot retrieve value; returning default";
+            LOG.warn(message);
+            return ReferenceWithError.newInstanceThrowingError(new BasicConfigValueAtContainer<TContainer,T>(getContainer(), ownKey, null, false,
+                    defaultValue),
+                new IllegalStateException(message));
+        }
+    }
+    
+    public List<ConfigValueAtContainer<TContainer,?>> getConfigAllInheritedRaw(ConfigKey<?> queryKey) {
```

Comment:
it's on the interface, useful for debugging and potentially for a config explorer, to know how a resolved value is derived.  agree we need test, will add.